Plugins: Refactor forward of cookies, OAuth token and header modifications by introducing client middlewares

marefr opened the pull request:
**What is this feature?**

Adding support for backend plugin client middlewares. Idea is to allow headers in outgoing `plugins.Client` and HTTP requests to be modified using client middlewares. 

The following client middlewares added:
- **Forward cookies:** Will forward incoming HTTP request Cookies to outgoing plugins.Client and HTTP requests if the datasource has enabled forwarding of cookies (keepCookies).
- **Forward OAuth token:** Will set OAuth token headers on outgoing plugins.Client and HTTP requests if the datasource has enabled Forward OAuth Identity (oauthPassThru).
- **Clear auth headers:** Will clear any outgoing HTTP headers that was part of the incoming HTTP request and used when authenticating to Grafana.

The current suggested way to register client middlewares is to have a separate package, `pluginsintegration`, responsible for  bootstrap/instantiate the backend plugin client with middlewares and/or longer term bootstrap/instantiate plugin management. Below some figures showing Grafana OSS and Enterprise could have different bootstrap/instantiation to allow potentially different middlewares in OSS and Enterprise as an example.

OSS:
![image](https://user-images.githubusercontent.com/1668778/199984161-b78b0efc-cb8b-41bf-909f-cd8479cd8dbd.png)

Enterprise:
![image](https://user-images.githubusercontent.com/1668778/199984195-47a4e370-000d-4bc3-81f4-9d1398a64705.png)

**Why do we need this feature?**

We shouldn't sprinkle the same/similar code in multiple different places around the code base. Better to implement it in one place, if possible.

**Who is this feature for?**

Plugin developers.

**Which issue(s) does this PR fix?**:
Fixes #54135
Related to #47734
Related to #57870
Related to #41623
Related to #57065

**Special notes for your reviewer**:
After discussion, decided to
- :heavy_check_mark: Move the send user header middleware and include docs for it to a separate PR to include that change in the changelog alone.
- :heavy_check_mark: Added backend plugin integration tests to validate headers are properly being forwarded to backend plugins and outgoing HTTP requests from plugins (core plugins)

marefr committed:
Plugins: Add support for client middleware

marefr commented on pkg/plugins/manager/client/forward_cookies_middleware.go:
```diff
@@ -0,0 +1,112 @@
+	reqCtx := contexthandler.FromContext(ctx)
+	// if request not for a datasource or no HTTP request context skip middleware
+	if req == nil || pCtx.DataSourceInstanceSettings == nil || reqCtx == nil || reqCtx.Req == nil {
+		return nil, nil
```
No you're right :+1:

marefr replied:
Pushed a fix

marefr committed:
fix after review

marefr committed:
add provider test

marefr committed:
ensure provider implements plugins.Client

marefr committed:
try use a new pluginsintegration service responsible for bootstraping plugin mgmt

marefr committed:
make it work with wire

marefr commented on pkg/services/pluginsintegration/pluginsintegration.go:
```diff
@@ -0,0 +1,30 @@
+)
+
+// WireSet provides a wire.ProviderSet of plugin providers.
+var WireSet = wire.NewSet(
```
Not sure it make sense to expose a new `wire.ProviderSet` since we don't do this anywhere else. The alternative would be to put the providers directly in wire.go as we normally do.

The bigger story here is that this `wire.ProviderSet` could provide all providers related to plugin management rather than keeping them in wire.go

marefr committed:
rename provider to decorator

marefr committed:
experiment move plugin providers from wire.go to pluginsintegration

marefr committed:
remove unneded aliases

marefr committed:
Merge remote-tracking branch 'origin/main' into poc_plugins_client_middleware

marefr commented on pkg/services/pluginsintegration/pluginsintegration.go:
```diff
@@ -0,0 +1,55 @@
+		clientmiddleware.NewForwardCookiesMiddleware(cfg),
+	}
+
+	return client.NewDecorator(c, clientMiddlewares...)
```
I guess a possible alternative doing this could be that the `client.Decorator` expose a method that allows registering new middlewares. Feels a bit more fragile and I think the current approach make things easier to understand. Thoughts?

xnyo replied:
I agree. I like this approach! :+1: Unless the middleware stack needs to change after initialization, I think this is great, and like you say it's cleaner and easier to understand

marefr committed:
clienttest package

marefr committed:
adds decorator tests

marefr committed:
forward cookies middleware tests

marefr committed:
move apply user header to proxyutil

marefr committed:
add tests and restructure a bit

marefr committed:
Merge remote-tracking branch 'origin/main' into poc_plugins_client_middleware

marefr committed:
add set and delete httpclient middleware

marefr committed:
Add user header client middleware

marefr committed:
Merge remote-tracking branch 'origin/main' into poc_plugins_client_middleware

marefr commented on pkg/services/pluginsintegration/clientmiddleware/cookies_middleware.go:
```diff
@@ -0,0 +1,121 @@
+		return ctx, nil
+	}
+
+	settings := pCtx.DataSourceInstanceSettings
+	jsonDataBytes, err := simplejson.NewJson(settings.JSONData)
+	if err != nil {
+		return ctx, err
+	}
+
+	ds := &datasources.DataSource{
+		Id:       settings.ID,
+		OrgId:    pCtx.OrgID,
+		JsonData: jsonDataBytes,
+		Updated:  settings.Updated,
+	}
```
JSON data bytes to struct in `DataSource`, back to bytes in `DataSourceInstanceSettings` and then back to `simplejson` is bad and feels like a hack, but can't figure out any other solution currently?

marefr commented on pkg/services/pluginsintegration/clientmiddleware/oauthtoken_middleware.go:
```diff
@@ -0,0 +1,135 @@
+		return ctx, nil
+	}
+
+	settings := pCtx.DataSourceInstanceSettings
+	jsonDataBytes, err := simplejson.NewJson(settings.JSONData)
+	if err != nil {
+		return ctx, err
+	}
+
+	ds := &datasources.DataSource{
+		Id:       settings.ID,
+		OrgId:    pCtx.OrgID,
+		JsonData: jsonDataBytes,
+		Updated:  settings.Updated,
+	}
```
JSON data bytes to struct in `DataSource`, back to bytes in `DataSourceInstanceSettings` and then back to `simplejson` is bad and feels like a hack, but can't figure out any other solution currently?

marefr committed:
Add oauthtoken middleware tests

marefr committed:
Document package and middlewares

marefr committed:
Add client middleware clear/remove auth headers used from incoming request when authentication to Grafana

marefr committed:
Remove code spread out for forwarding cookies and oauth token and clearing auth headers

marefr committed:
Update CODEOWNERS

marefr committed:
Merge remote-tracking branch 'origin/main' into poc_plugins_client_middleware

marefr committed:
spelling correction

marefr committed:
Merge remote-tracking branch 'origin/main' into poc_plugins_client_middleware

marefr committed:
Remove user header middleware

marefr committed:
Revert user header changes

wbrowne commented:
Apologies I forgot to submit this review 🤦 Will add the comment relating to the forward cookies on that PR

marefr commented on pkg/cmd/grafana-cli/runner/wire.go:
```diff
@@ -182,28 +175,13 @@ var wireSet = wire.NewSet(
 	uss.ProvideService,
-	pluginsCfg.ProvideConfig,
 	registry.ProvideService,
 	wire.Bind(new(registry.Service), new(*registry.InMemory)),
```
I might have missed those 😀

marefr commented on pkg/infra/httpclient/httpclientprovider/delete_headers_middleware_test.go:
```diff
@@ -0,0 +1,64 @@
+		require.ElementsMatch(t, []string{"final"}, ctx.callChain)
+
+		require.Equal(t, "a", req.Header.Get("X-Header-A"))
+		require.Equal(t, "d", req.Header.Get("X-Header-D"))
```
Good catch will fix

marefr committed:
wire/deps changes after review

marefr committed:
fixes after review

marefr commented on pkg/server/wireexts_oss.go:
```diff
@@ -8,8 +8,6 @@ import (
 
 	"github.com/grafana/grafana/pkg/models"
 	"github.com/grafana/grafana/pkg/plugins"
```
This one still here due to below

```go
wire.Bind(new(plugins.RoleRegistry), new(*acimpl.Service)),
```

I think it would make sense to have the implementation of plugins.RoleRegistry in the pluginsintegration package rather than in accesscontrol since it's an integration point between the two. Might be something to pick up later or what do you think?

marefr committed:
add backend plugin integration tests

marefr committed:
use testinfra.CreateUser

marefr committed:
Merge remote-tracking branch 'origin/main' into poc_plugins_client_middleware

marefr committed:
add forward oauth token integration tests

marefr committed:
Merge remote-tracking branch 'origin/main' into poc_plugins_client_middleware

marefr committed:
Merge remote-tracking branch 'origin/main' into poc_plugins_client_middleware

xnyo reviewed (changes requested):
Amazing work, very well done! 👏👏👏👏
So sorry if it took a while to review.
Just a few minor things/suggestions I have noticed that may need addressing:

marefr committed:
Merge remote-tracking branch 'origin/main' into poc_plugins_client_middleware

marefr commented on pkg/infra/httpclient/httpclientprovider/set_headers_middleware.go:
```diff
@@ -0,0 +1,29 @@
+// will be overwritten.
+func SetHeadersMiddleware(headers http.Header) httpclient.Middleware {
+	return httpclient.NamedMiddlewareFunc(SetHeadersMiddlewareName, func(opts httpclient.Options, next http.RoundTripper) http.RoundTripper {
+		if headers == nil {
```
Will fix

marefr commented on pkg/services/pluginsintegration/clientmiddleware/clear_auth_headers_middleware.go:
```diff
@@ -0,0 +1,87 @@
+	}
+
+	list := contexthandler.AuthHTTPHeaderListFromContext(ctx)
+	if list != nil {
```
It needs to be `len(list.Items)` in that case so I'll leave it as is.

marefr commented on pkg/infra/httpclient/httpclientprovider/set_headers_middleware.go:
```diff
@@ -0,0 +1,29 @@
+	"github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
+)
+
+const SetHeadersMiddlewareName = "forwarded-headers"
```
Good point. Will fix

marefr committed:
fixes after review

marefr committed:
improve clear auth headers tests

marefr commented on pkg/services/pluginsintegration/clientmiddleware/clear_auth_headers_middleware_test.go:
```diff
@@ -0,0 +1,288 @@
+			t.Run("Should attach delete headers middleware when calling QueryData", func(t *testing.T) {
+				_, err = cdt.Decorator.QueryData(req.Context(), &backend.QueryDataRequest{
+					PluginContext: pluginCtx,
+					Headers:       map[string]string{},
```
Good point. Rather than validating `QueryDataRequest.Headers` contains the extra added header I'll fix so that the HTTP request contains an extra header since that is where we delete headers from.

marefr replied:
After consideration populating `QueryDataRequest.Headers` etc as well just to be safe :+1:

marefr commented on pkg/services/pluginsintegration/clientmiddleware/cookies_middleware_test.go:
```diff
@@ -0,0 +1,167 @@
+		t.Run("Should not forward cookies when calling QueryData", func(t *testing.T) {
+			_, err = cdt.Decorator.QueryData(req.Context(), &backend.QueryDataRequest{
+				PluginContext: pluginCtx,
+				Headers:       map[string]string{},
```
I'll fix :+1:

marefr committed:
improve cookies middleware tests

marefr committed:
improve oauthtoken middleware tests

marefr commented on pkg/services/pluginsintegration/clientmiddleware/oauthtoken_middleware_test.go:
```diff
@@ -0,0 +1,158 @@
+		t.Run("Should forward OAuth Identity when calling QueryData", func(t *testing.T) {
+			_, err = cdt.Decorator.QueryData(req.Context(), &backend.QueryDataRequest{
+				PluginContext: pluginCtx,
+				Headers:       map[string]string{},
```
I'll fix :+1:

marefr replied:
I also refactored and removed the `ForwardedOAuthIdentityMiddleware` in favor of using `SetHeadersMiddleware`.

marefr committed:
remove ForwardedOAuthIdentityMiddleware in favor of SetHeadersMiddleware

xnyo reviewed (approved):
LGTM 🚀🚀 Amazing work!

wbrowne reviewed (approved):
I think it looks good! 😅 Took it for a spin and found no issues ✅ 

Awesome work throwing this all together 🚀

marefr committed:
Merge remote-tracking branch 'origin/main' into poc_plugins_client_middleware

marefr committed:
move header name references into constants

marefr committed:
Merge remote-tracking branch 'origin/main' into poc_plugins_client_middleware

marefr committed:
Merge remote-tracking branch 'origin/main' into poc_plugins_client_middleware

marefr committed:
rename query_parsing.go to models.go

marefr committed:
add test verifying user and orgID is forwarded to QueryData from expression requests

marefr committed:
extend backend plugin integration tests to include regular and expression query

In addition, adds some better input validation of plugin client and decorator,
making sure input arguments are not nil.

marefr committed:
Merge remote-tracking branch 'origin/main' into poc_plugins_client_middleware

marefr commented:
@wbrowne @xnyo updated to resolve conflicts caused by #58897. Basically reverting most of the changes in #58897 since this is targeting v9.4.x:
- Added a unit test to query service that verifies `User` and `OrgID` are forwarded to server side expressions (SSE) and that SSE are using these when calling `<plugin client>.QueryData`
- Extended the backend plugin integration tests to include /api/ds/query being called with both regular (non-expression) and expression queries.
- Added some better input validation of plugin client and client decorator making sure input arguments are not nil. It should never happen so good to have some tests of it since other unit/integration tests are automatically verifying this as well.

Hopefully you've marked the already reviewed files as viewed and should simplify the review of the latest changes. Otherwise, suggest looking at each commit.

wbrowne reviewed (approved):
LGTM 🚀 

Just added a comment as noticed something that makes sense but wanted to verify

xnyo reviewed (approved):
LGTM! :rocket:

marefr commented on pkg/services/pluginsintegration/clientmiddleware/utils_test.go:
```diff
@@ -0,0 +1,13 @@
+package clientmiddleware
```
Yeah I since only used for tests. I'll change this to testing.go or something

marefr committed:
rename file

marefr committed:
Merge remote-tracking branch 'origin/main' into poc_plugins_client_middleware
